internal/linker/link_node_modules.ts

```diff
@@ -276,36 +297,72 @@ export async function main(args: string[], runfiles: Runfiles) {
   // Now add symlinks to each of our first-party packages so they appear under the node_modules tree
   const links = [];
 
-  const linkModule =
-      async (name: string, root: LinkerRoot, modulePath: string) => {
-    let target: string = '<package linking failed>';
-    switch (root) {
-      case 'bin':
-        // FIXME(#1196)
-        target = path.join(workspaceAbs, bin, toWorkspaceDir(modulePath));
-        break;
-      case 'src':
-        target = path.join(workspaceAbs, toWorkspaceDir(modulePath));
-        break;
-      case 'runfiles':
-        target = runfiles.resolve(modulePath) || '<runfiles resolution failed>';
-        break;
-    }
+  const linkModules =
+      async (modules: {name: string, root: LinkerRoot, modulePath: string}[]) => {
+    for (const m of modules) {
+      let target: string = '<package linking failed>';
+      switch (m.root) {
+        case 'bin':
+          // FIXME(#1196)
+          target = path.join(workspaceAbs, bin, toWorkspaceDir(m.modulePath));
+          break;
+        case 'src':
+          target = path.join(workspaceAbs, toWorkspaceDir(m.modulePath));
+          break;
+        case 'runfiles':
+          target = runfiles.resolve(m.modulePath) || '<runfiles resolution failed>';
+          break;
+      }
 
-    await symlink(target, name);
+      // ensure the subdirectories exist
+      mkdirp(path.dirname(m.name));
+      await symlink(target, m.name);
+    }
   }
 
-  for (const m of Object.keys(modules)) {
-    const segments = m.split('/');
-    if (segments.length > 2) {
-      throw new Error(`module ${m} has more than 2 segments which is not a valid node module name`);
-    }
-    if (segments.length == 2) {
-      // ensure the scope exists
-      mkdirp(segments[0]);
-    }
-    const [kind, modulePath] = modules[m];
-    links.push(linkModule(m, kind, modulePath));
+  // Group nested modules as these need to be symlinked in order.
+  // For example, given a list of module keys such as:
+  // ['a', '@foo/c/c/c/c', 'b/b', 'b', '@foo/c', '@foo/c/c']
+  // this reducer should output the groups list:
+  // [ [ '@foo/c', '@foo/c/c', '@foo/c/c/c/c' ], [ 'a' ], [ 'b', 'b/b' ] ]
+  const groupedModules = Object.keys(modules).sort().reduce(
+      (grouped: string[][], module: string, index: number, array: string[]): string[][] => {
+        if (index > 0 && module.startsWith(`${array[index - 1]}/`)) {
+          grouped[grouped.length - 1].push(module);
+        } else {
+          grouped.push([module]);
+        }
+        return grouped;
+      }, []);
+
+  log_verbose(`grouped modules ${JSON.stringify(groupedModules)}`);
+  for (const group of groupedModules) {
+    // reduce links such as `@foo/b/c => /path/to/a/b/c` to their
```

Comment:
added this after discussing with @alexeagle night. If there is a `@foo/a/b/c/d` module mapping in your deps and it maps to a path `/path/to/foo/a/b/c/d`, for example, then the mapping can and should be reduced to `@foo` => `/path/to/foo`.